File: manual/c/arrays.c
#include "arrays.h"

#include <stdio.h>

/*
 * Global variables
 */

int arr2[3] = {2, 3, 4};

int arr3[] = {3, 4, 5};


int sudoku[3][3] = { {9, 8, 7}
                   , {6, 5, 4}
                   , {3, 2, 1}
                   };

int triplets[][3] = { {0, 0, 0}
                    , {1, 1, 1}
                    };

triplet baz = {9,1,8};

/*
 * Matrix transpose
 */

void transpose (matrix input, matrix output) {
  int i, j;
  for (i = 0; i < 3; i++)
      for (j = 0; j < 3; j++)
          output[i][j] = input[j][i];
};

/*
 * Complex examples
 */

int (*global_triplet_ptrs[])[3] = { &triplets[0], &triplets[1], &triplets[0] };

void pretty_print_triplets (int (*ts[])[3]) {
  // Pointer to the first triplet
  triplet *t1 = ts[0];
  // Pointer to the second triplet
  triplet *t2 = ts[1];
  // Pointer to the third triplet
  triplet *t3 = ts[2];
  printf("{%d, %d, %d}\n", (*t1)[0], (*t1)[1], (*t1)[2]);
  printf("{%d, %d, %d}\n", (*t2)[0], (*t2)[1], (*t2)[2]);
  printf("{%d, %d, %d}\n", (*t3)[0], (*t3)[1], (*t3)[2]);
}

File: manual/hs/manual/app/Manual/Types/Arrays.hs
{-# LANGUAGE DataKinds #-}

module Manual.Types.Arrays (examples) where

import Foreign as F
import GHC.TypeNats

import HsBindgen.Runtime.ConstantArray qualified as CA
import HsBindgen.Runtime.ConstPtr
import HsBindgen.Runtime.IncompleteArray qualified as IA

import Arrays qualified
import Arrays.Global qualified as Arrays
import Arrays.Safe qualified as Arrays
import Manual.Tools

{-------------------------------------------------------------------------------
  Array utilities
-------------------------------------------------------------------------------}

reverseConstantArray :: (Storable a, Show a, KnownNat n) => Ptr (CA.ConstantArray n a) -> IO ()
reverseConstantArray ptr = do
    -- Print the input contents
    xs <- F.peek ptr
    print xs
    -- Reverse the array
    let ys = CA.fromList . reverse . CA.toList $ xs
    F.poke ptr ys
    -- Print the output contents
    zs <- F.peek ptr
    print zs

reverseConstantArrayElems :: (Storable a, Show a, KnownNat n) => Ptr (CA.ConstantArray n a) -> IO ()
reverseConstantArrayElems ptr = do
    let (p, ptr') = CA.toFirstElemPtr ptr
    -- Print the input contents
    xs <- F.peekArray (CA.intVal p) ptr'
    print xs
    -- Reverse the array
    let ys = reverse xs
    F.pokeArray ptr' ys
    -- Print the output contents
    zs <- F.peekArray (CA.intVal p) ptr'
    print zs

reverseIncompleteArray :: (Storable a, Show a) => Int -> Ptr (IA.IncompleteArray a) -> IO ()
reverseIncompleteArray n ptr = do
    -- Print the input contents
    xs <- IA.peekArray n ptr
    print xs
    -- Reverse the array
    let ys = IA.fromList . reverse . IA.toList $ xs
    IA.pokeArray ptr ys
    -- Print the output contents
    zs <- IA.peekArray n ptr
    print zs

reverseIncompleteArrayElems :: (Storable a, Show a) => Int -> Ptr (IA.IncompleteArray a) -> IO ()
reverseIncompleteArrayElems n ptr = do
    let ptr' = IA.toFirstElemPtr ptr
    -- Print the input contents
    xs <- F.peekArray n ptr'
    print xs
    -- Reverse the array
    let ys = reverse xs
    F.pokeArray ptr' ys
    -- Print the output contents
    zs <- F.peekArray n ptr'
    print zs

transposeMatrix :: Arrays.Matrix -> IO Arrays.Matrix
transposeMatrix inputMatrix =
    CA.withPtr inputMatrix $ \inputPtr -> do
      F.alloca $ \(outputPtr :: Ptr Arrays.Matrix) -> do
        Arrays.transpose_wrapper (ConstPtr inputPtr) (snd $ CA.toFirstElemPtr outputPtr)
        peek outputPtr

{-------------------------------------------------------------------------------
  Examples
-------------------------------------------------------------------------------}

examples :: IO ()
examples = do
    section "Arrays"

    subsection "Global variables"
    reverseConstantArray Arrays.arr1
    reverseConstantArrayElems Arrays.arr1

    reverseConstantArray Arrays.arr2
    reverseConstantArrayElems Arrays.arr2

    reverseIncompleteArray 3 Arrays.arr3
    reverseIncompleteArrayElems 3 Arrays.arr3

    print =<< F.peek Arrays.sudoku
    print =<< IA.peekArray 2 Arrays.triplets

    subsection "Matrix transpose"
    let inputMatrix = Arrays.Matrix $
          CA.fromList [
              Arrays.Triplet $ CA.fromList [1, 2, 3]
            , Arrays.Triplet $ CA.fromList [4, 5, 6]
            , Arrays.Triplet $ CA.fromList [7, 8, 9]
            ]
    print inputMatrix
    outputMatrix <- transposeMatrix inputMatrix
    print outputMatrix

    subsection "Complex example"
    ts <- IA.peekArray 2 Arrays.triplets
    let tripletAddresses = [advancePtr (IA.toFirstElemPtr Arrays.triplets) n | n <- [0..]]
    print (zip (IA.toList ts) tripletAddresses)
    print =<< IA.peekArray 3 Arrays.global_triplet_ptrs
    Arrays.pretty_print_triplets (castPtr Arrays.global_triplet_ptrs)
